Lucid/Utils/AsyncCurrentValue.swift

```diff
@@ -0,0 +1,137 @@
+//
+//  AsyncCurrentValue.swift
+//  Lucid
+//
+//  Created by Stephane Magne on 2023-05-15.
+//  Copyright © 2023 Scribd. All rights reserved.
+//
+
+import Foundation
+
+public protocol AsyncCurrentValueDelegate: AnyObject {
+
+    func didRemoveFinalIterator() async
+}
+
+/**
+    This class is meant to replicate Combine.CurrentValueSubject.
+
+    The use case looks like this...
+
+    In the calling class:
+
+        Task {
+            let currentValue = manager.currentValue
+            for await value in currentValue where Task.isCancelled == false {
+                // perform some action
+            }
+        }
+
+     In the class owning AsyncCurrentValue:
+
+        currentValue.update(newValue)
+*/
+public final class AsyncCurrentValue<Element>: AsyncSequence {
+
+    private(set) var value: Element
+
+    private var repository = AsyncCurrentValueRepository<Element>()
+
+    public init(_ initialValue: Element) {
+        self.value = initialValue
+    }
+
+    public func makeAsyncIterator() -> AsyncCurrentValueIterator<Element> {
+        let iterator = AsyncCurrentValueIterator<Element>(value: value)
+        Task {
+            await repository.addIterator(iterator)
```

Comment:
One iterator is created per observer.

Any time you create a for loop, you need to pass out a new iterator to that call-site.

```swift
for try await value in currentValue { 
     // action
} 
```

Otherwise, they all iterate through the same array and each only get partial information.

Reply:
Understood